src/coreclr/jit/valuenum.cpp

```diff
@@ -5298,7 +5298,9 @@ ValueNum ValueNumStore::VNForLoad(ValueNumKind vnk,
     // is the same as its VN's).
     loadValue = VNForLoadStoreBitCast(loadValue, loadType, loadSize);
 
-    assert(genActualType(TypeOfVN(loadValue)) == genActualType(loadType));
+    // NOTE: it's allowed to bitcast TYP_REF to TYP_I_IMPL for nongc handles.
+    assert((genActualType(TypeOfVN(loadValue)) == genActualType(loadType)) ||
+           ((IsVNObjHandle(loadValue) && (loadType == TYP_REF))));
```

Comment:
This is not right, it breaks the invariant we're trying to preserve here. The right fix is to create `TYP_REF`-typed `VNForHandle`s for frozen objects.

Reply:
Any idea how to do it? Aren't all other places assigning TYP_LONG ObjHandle VN to TYP_REF trees doing the same?